packages/cli/src/util/get-latest-version/get-latest-worker.js

```diff
@@ -14,36 +14,101 @@ const fetch = require('node-fetch');
 const path = require('path');
 const { Agent: HttpsAgent } = require('https');
+const { bold, gray, red } = require('chalk');
+const { format, inspect } = require('util');
 
-process.on('unhandledRejection', err => {
-  if (process.connected) {
-    console.error('Exiting update worker due to error:');
-    console.error(err);
+/**
+ * An simple output helper which accumulates error and debug log messages in
+ * memory for potential persistance to disk while immediately outputting errors
+ * and debug messages, when the `--debug` flag is set, to `stderr`.
+ */
+class WorkerOutput {
+  debugLog = [];
+  logFile = null;
+
+  constructor({ debug = true }) {
+    this.debugOutputEnabled = debug;
+  }
+
+  debug(...args) {
+    this.print('debug', args);
+  }
+
+  error(...args) {
+    this.print('error', args);
+  }
+
+  print(type, args) {
+    const str = format(
+      ...args.map(s => (typeof s === 'string' ? s : inspect(s)))
+    );
+    this.debugLog.push(`[${new Date().toISOString()}] [${type}] ${str}`);
+    if (type === 'debug' && this.debugOutputEnabled) {
+      console.error(
+        `${gray('>')} ${bold('[debug]')} ${gray(
+          `[${new Date().toISOString()}]`
+        )} ${str}`
+      );
+    } else if (type === 'error') {
+      console.error(`${red(`Error:`)} ${str}`);
+    }
   }
+
+  setLogFile(file) {
+    if (this.logFile === null) {
+      process.on('exit', () => {
+        if (this.debugLog.length) {
+          fs.outputFileSync(this.logFile, this.debugLog.join('\n'));
+        }
+      });
+    }
+    this.logFile = file;
+  }
+}
+
+const output = new WorkerOutput({
+  debug: process.argv.includes('--debug') || !process.connected,
+});
+
+process.on('unhandledRejection', err => {
+  output.error('Exiting worker due to error:', err);
   process.exit(1);
 });
 
 const defaultGetLatestInterval = 1000 * 60 * 60 * 24 * 7; // 1 week
 
 // this timer will prevent this worker process from running longer than 10s
-const timer = setTimeout(() => process.exit(1), 10000);
+const timer = setTimeout(() => {
+  output.error('Worker timed out after 10 seconds');
+  process.exit(1);
+}, 10000);
 
 // wait for the parent to give us the work payload
 process.once('message', async msg => {
+  output.debug('Received message from parent:', msg);
+
+  output.debug('Disconnecting from parent');
   process.disconnect();
 
   const { cacheFile, distTag, name, getLatestInterval } = msg;
   const cacheFileParsed = path.parse(cacheFile);
   await fs.mkdirp(cacheFileParsed.dir);
 
+  output.setLogFile(
+    path.join(cacheFileParsed.dir, `${cacheFileParsed.name}.log`)
+  );
+
   // check for a lock file and either bail if running or write our pid and continue
   const lockFile = path.join(
     cacheFileParsed.dir,
     `${cacheFileParsed.name}.lock`
   );
+  output.debug(`Checking lock file: ${lockFile}`);
   if (await isRunning(lockFile)) {
-    return;
+    output.debug('Worker already running, exiting');
+    process.exit(2);
```

Comment:
**question (non-blocking):** Does this need to exit instead of return?

**suggestion (non-blocking):** change exit code to `1`

In a team meeting, we aligned on using these exit codes.

- 0: success
- 1: internal system error
- 2: help output
- 4: user error
- 5: external system error (ex: API Error)

While not official, currently followed, or enforced, if we want to start moving towards this convention, we can start here by changing the exit code of `2` to `1`.

Reply:
Yes, we need an exit. I found that without an explicit exit here, the 10 second time was firing despite clearing the timer. I'm not sure what exactly was keeping the process from exiting, but this "already running" check was added after I realized I needed the exit calls.

Good call on the exit code. Fixed!